### PR TITLE
Sven update

### PR DESCRIPTION
**Pull request by AaronDavidSchneider**

### CODE REVIEW

**Review by robinbaeyens — DISMISSED** [review]
Nice writing class, it looks very professional. I also approve the better grouping of functions.
I just corrected and committed a few typo's.
I also found a few small inconsistencies that I think you should check first, but overall it looks good!

**Comment by AaronDavidSchneider** [comment]
@Kiefersv: Linking you here in case you missed @robinbaeyens comments

File: GCMtools/GCMtools.py [hunk 1]
```diff
@@ -108,16 +98,40 @@ def get_models(self, tag=None):
...
-        # If no tag is given, return all models
-        if tag is None:
-            return self.models
-        # If the tag is valid, return the corresponding model
-        elif isinstance(tag, str):
-            return self._models[tag]
-        # If the tag is not a string, raise an error
-        else:
-            raise ValueError('The given tag is not a string.')
+        return self._models.get_models()
```
[the remark this answers is not in the document]
Comment (Kiefersv) [reply GCMtools/GCMtools.py:101]: absolutly

File: GCMtools/utils/manipulations.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    TODO
+    """
+    # print information
+    wrt.write_status('STAT', 'Calculate horizontal average')
+    wrt.write_status('INFO', 'Tag: ' + tag)
+    wrt.write_status('INFO', 'Varaible to be plotted: ' + var_key)
+    wrt.write_status('INFO', 'Output variable: ' + var_key_out)
+    wrt.write_status('INFO', 'Area of grid cells: ' + area_key)
+
+    ds = self._get_one_model(tag)
```
[the remark this answers is not in the document]
Comment (Kiefersv) [reply GCMtools/utils/manipulations.py:36]: thanks, I missed that one

File: GCMtools/GCMtools.py [hunk 3]
```diff
@@ -364,12 +256,13 @@ def isobaric_slice(self, var_key, p, tag=None, **kwargs):
             and multiple datasets are available, an error is raised.
         """
         # select the appropriate dataset
-        ds = self._get_one_model(tag)
+        ds = self._models.get_one_model(tag)
```
[the remark this answers is not in the document]
Comment (AaronDavidSchneider) [reply GCMtools/GCMtools.py:275]: I think I agree with Sven that its a good idea to have the security layer here and use the underlying classes structure.

**Comment by AaronDavidSchneider** [comment]
Hi @robinbaeyens, @Kiefersv,

I cleaned up the PR by rebasing on the main branch, readding the `models` property for the GCMT class (for quick access).

The following things are still needed (IMO):
- [ ] testing (havent yet used it and tested it)
- [ ] checking that the docs still compile

I guess we also need to discuss if we want to use @Kiefersv's IO Class, or instead just use pythons inbuilt `logging` package (https://docs.python.org/3/howto/logging.html).

Thanks @Kiefersv for your work! I hope I did not break your changes with my rebase :D

Cheers
Aaron

File: GCMtools/utils/manipulations.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        and multiple datasets are available, an error is raised.
+    var_key_out: str, optional
+        variable name used to store the outcome. If not provided, this script will just
+        return the overturning circulation and not change the dataset inplace.
+    """
+
+    # print information
+    wrt.write_status('STAT', 'Calculate meridional overturning streamfunction')
+    wrt.write_status('INFO', 'Tag: ' + tag)
+    wrt.write_status('INFO', 'Output variable: ' + var_key_out)
```
Comment (AaronDavidSchneider) [inline GCMtools/utils/manipulations.py:74]: These lines will fail, since `wrt` is not known.

Comment (Kiefersv) [reply GCMtools/utils/manipulations.py:74]: Thx! Will change it right away

**Comment by Kiefersv** [comment]
Hey All,

Thanks again for looking at my changes in detail!

I tested the rebase and it works fine for my small example but you are right, it would be best if the two of you could test it on your examples as well (as Aaron is apparently already doing while I write this message). On the same note, it might be worth thinking about unit tests at some point.

Regarding the I/O class, I have no strong opinion on whether to use mine or a package. I did not have time to read into the Logging package but will do at some point. Feel free to change it if anyone of you has time to look into it before I do.

**Review by AaronDavidSchneider — COMMENTED** [review]
Ok. I did thorough tests and checked the docs. There were still a few bugs like models not being returned and stuff.
From my side, this PR is good to go for now.

Future improvements would be (like @Kiefersv said) adding tests (!!!) and maybe a CI that automatically checks these tests. I propose that we stick to the writer class that @Kiefersv proposes and change it to something different at later times.

So thanks again @Kiefersv!

If both of you @Kiefersv and @robinbaeyens could give a final approve or comments, I will merge the PR.

**Pull request merged**